### PR TITLE
[Application] Add Istio Bookinfo application in OAM way

### PR DESCRIPTION
**Pull request by zzxwill**
This application demonstrates how OAM bootstraps [Istio Bookinfo application](https://istio.io/latest/docs/examples/bookinfo/)
with workload `Gateway`, `VirtualService`, `DestinationRule` deployed by OAM.

### Checklist

I have:
- [ ] Title of the PR starts with OAM type (e.g. `[Workload]` or `[Trait]` or `[Scope]`).
- [ ] Updated/Added any relevant [documentation] and [examples].
- [ ] Unit/E2E Tests added.

### CODE REVIEW

**Comment by zzxwill** [comment]
Please help take a look at those four TODO in `README.me`, thanks.

File: applications/istio/bookinfo/README.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+```
+$ kubectl apply -f https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml
+workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created
+traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev created
+workloaddefinition.core.oam.dev/postgresqlinstancerequirements.database.example.org created
+
+```
+
+(TODO)  ** The root cause why this depends on installation of Crossplane has to be investigated.**
```
Comment (szihai) [inline applications/istio/bookinfo/README.md:26]: If this issue cannot be solved in recent time, just ask users to install crossplane and avoid the confusion

File: applications/istio/bookinfo/README.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+$ kubectl apply -f https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml
+
+unable to recognize "https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml": no matches for kind "WorkloadDefinition" in version "core.oam.dev/v1alpha2"
+unable to recognize "https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml": no matches for kind "TraitDefinition" in version "core.oam.dev/v1alpha2"
+unable to recognize "https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml": no matches for kind "WorkloadDefinition" in version "core.oam.dev/v1alpha2"
+```
+
+- Istio 1.6.3
+
+Pleaser refer to [Download and Install Istio](https://istio.io/latest/docs/setup/getting-started/).
```
Comment (szihai) [inline applications/istio/bookinfo/README.md:37]: add "And do not deploy the book info sample from the istio installation folder".

File: applications/istio/bookinfo/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+## Deploy the application
+
+```
+$ kubectl apply -f bookinfo.yaml
+```
+(TODO) Not sure whether `Service` and `Deployment` from [bookinfo.yaml](./bookinfo.yaml) should be converted to
```
Comment (szihai) [inline applications/istio/bookinfo/README.md:44]: I would suggest to use the components. And please remove the bookinfo.yaml.

Comment (zzxwill) [reply applications/istio/bookinfo/README.md:44]: Is this way, the appconfig file will become so long. Still hesitating.

**Comment by szihai** [comment]
Followed steps, get into 2 errors:
a. `k get VirtualService bookinfo -o yaml`   Error from server (NotFound): virtualservices.networking.istio.io "bookinfo" not found. The same with `k get Gateway bookinfo-gateway` command.
b. Got the ingress provisioned. `istio-ingressgateway   LoadBalancer   172.21.1.228   47.251.32.12   15021:30957/TCP,80:31562/TCP,443:32107/TCP,15443:30771/TCP`. But when trying `http://47.251.32.12/productpage`, got "connection refused" error.

File: applications/istio/bookinfo/README.md (hunk 3 above)
Comment (resouer) [reply applications/istio/bookinfo/README.md:44]: Service is trait and Deployment is workload, let's keep everything straightforward (i.e. use OAM as k8s app definition only) for the first step and think about abstractions (`ContainerizedWorkload` etc) later. For abstractions, it should be more than a simple ContainerizedWorkload - new Trait and Scope abstractions are needed. 

@zzxwill In the context above, you don't even need to install `addon-oam-kubernetes-local`. But you need to define/generate definition objects for your workload and trait.

**Comment by zzxwill** [comment]
> Followed steps, get into 2 errors:
> a. `k get VirtualService bookinfo -o yaml` Error from server (NotFound): virtualservices.networking.istio.io "bookinfo" not found. The same with `k get Gateway bookinfo-gateway` command.
> b. Got the ingress provisioned. `istio-ingressgateway LoadBalancer 172.21.1.228 47.251.32.12 15021:30957/TCP,80:31562/TCP,443:32107/TCP,15443:30771/TCP`. But when trying `http://47.251.32.12/productpage`, got "connection refused" error.

Have you tried this step at https://istio.io/latest/docs/setup/getting-started/#ip?
```
$ kubectl apply -f samples/bookinfo/networking/bookinfo-gateway.yaml
gateway.networking.istio.io/bookinfo-gateway created
virtualservice.networking.istio.io/bookinfo created
```

File: applications/istio/bookinfo/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+```
+kubectl create namespace crossplane-system
+helm repo add crossplane-alpha https://charts.crossplane.io/alpha
+helm install addon-oam-kubernetes-local --namespace crossplane-system crossplane-alpha/oam-core-resources
+```
+
+- OAM workloads/traits definitions
+
+```
+$ kubectl apply -f https://raw.githubusercontent.com/crossplane/crossplane/release-0.12/docs/snippets/run/definitions.yaml
```
[the remark this answers is not in the document]
Comment (szihai) [reply applications/istio/bookinfo/README.md:19]: No we don't.

File: applications/istio/bookinfo/README.md (hunk 1 above)
Comment (zzxwill) [reply applications/istio/bookinfo/README.md:26]: @ryanzhang-oss If we don't install Crossplane (`helm install crossplane --namespace crossplane-system crossplane-alpha/crossplane`), we will hit issue as below when applying  `WorkloadDefinition` and `TraitDefinition` manifests.

**Comment by szihai** [comment]
Now it works!

File: applications/istio/bookinfo/appconfig-2-bookinfo-DestinationRule.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: bookinfo
+spec:
+  components:
+    - componentName: bookinfo-gateway
+      traits:
+        - trait:
+            apiVersion: networking.istio.io/v1alpha3
```
Comment (CoderPoet) [inline applications/istio/bookinfo/appconfig-2-bookinfo-DestinationRule.yaml:10]: Need to add add TraitDefinition for istio CRDs, for example
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: virtualservices.networking.istio.io
spec:
  definitionRef:
    name: virtualservices.networking.istio.io
---    
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: destinationrules.networking.istio.io
spec:
  definitionRef:
    name: destinationrules.networking.istio.io
```

File: applications/istio/bookinfo/bookinfo.yaml [hunk 6]
```diff
@@ -0,0 +1,331 @@
+# Copyright 2017 Istio Authors
```
Comment (CoderPoet) [inline applications/istio/bookinfo/bookinfo.yaml:1]: The creation of native resources should not be needed, and should be componentized

File: applications/istio/bookinfo/appconfig-1-bookinfo-VirtualService.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: bookinfo
+spec:
+  components:
+    - componentName: bookinfo-gateway
```
Comment (CoderPoet) [inline applications/istio/bookinfo/appconfig-1-bookinfo-VirtualService.yaml:7]: Here you should need to reference the previously created components, for example

```yaml
    - componentName: reviews
    - componentName: details
    - componentName: ratings
    - componentName: productpage
    - componentName: reviews-v1
    - componentName: reviews-v2
    - componentName: reviews-v3
    - componentName: details-v1    
    - componentName: ratings-v1    
    - componentName: productpage-v1
    - componentName: bookinfo-gateway
```

In addition, I think service should also be used as a deployment trait, not a component

**Comment by szihai** [comment]
Add "kubectl apply -f Scopes/" explicitly in the README.
Add a cleanup script.

**Comment by zzxwill** [comment]
> Here are my 2 suggestions:
> Add "kubectl apply -f Scopes/" explicitly in the README.
> Add a cleanup instruction.
> 
> Other than those, everything is clear and works fine for me.

Thanks a lot, cleanup is essential nearly for every demo.

**Pull request merged**